### PR TITLE
Callback pagination

### PR DESCRIPTION
**Pull request by athlan**
Since there is [pagination support](https://github.com/KnpLabs/knp-components/tree/master/src/Knp/Component/Pager/Event/Subscriber/Paginate) for:
* Arrays
* Elastica Query
* Proper
* Solarium
* Doctrine

Nowadays hexagonal architectures and Domain-Driven Design becomes more and more popular. This implicates to have some abstraction to repositories of data. It's code quality breach to expose implementation details and have pagination over it.

The proposition is to implement an interface being able to paginate over custom callback functions (one for get count, one for get items by limit/offset).

Note: Please squash merge.

Example usage:

```php
$count = function () {
    // return count from repository
};
$items = function ($offset, $limit) {
    // return data from repository using offet and limit
};
$data = CallbackPagination($count, $items);

// pagination here
$pagination = $p->paginate($data, $page, $itemsPerPage);
```

### CODE REVIEW

File: src/Knp/Component/Pager/Event/Subscriber/Paginate/Callback/CallbackPagination.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        return call_user_func($this->count);
+    }
+
+    /**
+     * @param $offset
+     * @param $limit
+     * @return array
+     */
+    public function getPaginationItems($offset, $limit): array
```
Comment (garak) [inline src/Knp/Component/Pager/Event/Subscriber/Paginate/Callback/CallbackPagination.php]: What about forcing types? If so, we can remove phpdoc comments, where unneeded

File: src/Knp/Component/Pager/Event/Subscriber/Paginate/Callback/CallbackSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Callback pagination.
+ *
+ * @author Piotr Pelczar <me@athlan.pl>
+ */
+class CallbackSubscriber implements EventSubscriberInterface
+{
+    public function items(ItemsEvent $event)
```
Comment (garak) [inline src/Knp/Component/Pager/Event/Subscriber/Paginate/Callback/CallbackSubscriber.php]: return type here

File: src/Knp/Component/Pager/Event/Subscriber/Paginate/Callback/CallbackSubscriber.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+            }
+            else {
+                $event->items = [];
+            }
+
+            $event->stopPropagation();
+        }
+    }
+
+    public static function getSubscribedEvents()
```
Comment (garak) [inline src/Knp/Component/Pager/Event/Subscriber/Paginate/Callback/CallbackSubscriber.php]: return type

File: tests/Test/Pager/Subscriber/Paginate/Callback/CallbackTest.php [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+use Knp\Component\Pager\Event\Subscriber\Paginate\Callback\CallbackPagination;
+use Knp\Component\Pager\Event\Subscriber\Paginate\Callback\CallbackSubscriber;
+use Test\Tool\BaseTestCase;
+use Knp\Component\Pager\Paginator;
+use Symfony\Component\EventDispatcher\EventDispatcher;
+use Test\Mock\PaginationSubscriber as MockPaginationSubscriber;
+use Knp\Component\Pager\Pagination\PaginationInterface;
+
+class CallbackTest extends BaseTestCase
```
Comment (garak) [inline tests/Test/Pager/Subscriber/Paginate/Callback/CallbackTest.php]: please make class final

**Comment by garak** [comment]
I think that this PR should also include some documentations. I know that docs are, currently, a bit lacking, but for sure we should start somehow ;-)

File: src/Knp/Component/Pager/Event/Subscriber/Paginate/Callback/CallbackSubscriber.php (hunk 3 above)
Comment (athlan) [reply src/Knp/Component/Pager/Event/Subscriber/Paginate/Callback/CallbackSubscriber.php]: fixed

File: src/Knp/Component/Pager/Event/Subscriber/Paginate/Callback/CallbackSubscriber.php (hunk 2 above)
Comment (athlan) [reply src/Knp/Component/Pager/Event/Subscriber/Paginate/Callback/CallbackSubscriber.php]: fixed

File: src/Knp/Component/Pager/Event/Subscriber/Paginate/Callback/CallbackPagination.php (hunk 1 above)
Comment (athlan) [reply src/Knp/Component/Pager/Event/Subscriber/Paginate/Callback/CallbackPagination.php]: Good point, fixed

File: tests/Test/Pager/Subscriber/Paginate/Callback/CallbackTest.php (hunk 4 above)
Comment (athlan) [reply tests/Test/Pager/Subscriber/Paginate/Callback/CallbackTest.php]: none of test suite is final now, it would be inconsequential

Comment (garak) [reply tests/Test/Pager/Subscriber/Paginate/Callback/CallbackTest.php]: We need to start somewhere...

Comment (athlan) [reply tests/Test/Pager/Subscriber/Paginate/Callback/CallbackTest.php]: all right, introduced

**Comment by athlan** [comment]
Agree regarding documentation. I've put some notes in `intro.md` right now, however docs should be structured somehow (general usage, pagination types, supported targets).

CR fixes applied.

File: src/Knp/Component/Pager/Event/Subscriber/Paginate/Callback/CallbackPagination.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    public function getPaginationCount(): int
+    {
+        return call_user_func($this->count);
+    }
+
+    public function getPaginationItems(int $offset, int $limit): array
+    {
+        return call_user_func_array($this->items, array($offset, $limit));
```
Comment (garak) [inline src/Knp/Component/Pager/Event/Subscriber/Paginate/Callback/CallbackPagination.php]: Please use short array syntax (not just here)

Comment (athlan) [reply src/Knp/Component/Pager/Event/Subscriber/Paginate/Callback/CallbackPagination.php]: fixed

File: src/Knp/Component/Pager/Event/Subscriber/Paginate/Callback/CallbackSubscriber.php [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+                $event->items = [];
+            }
+
+            $event->stopPropagation();
+        }
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return array(
```
Comment (garak) [inline src/Knp/Component/Pager/Event/Subscriber/Paginate/Callback/CallbackSubscriber.php]: this

Comment (athlan) [reply src/Knp/Component/Pager/Event/Subscriber/Paginate/Callback/CallbackSubscriber.php]: scatty today, fixed

**Pull request merged**